### ISSUE TITLE
Contacts not showing up

### ISSUE THREAD

**Issue by rannbaron** [signalapp/Signal-Android#7394]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

----------------------------------------

### Bug description
Contacts not displaying.

### Steps to reproduce
- Install Signal

**Actual result:** Describe here what happens after you run the steps above (i.e. the buggy behaviour)

No contact names were displayed.

**Expected result:** Describe here what should happen after you run the steps above (i.e. what would be the correct behaviour)

See contacts.

### Device info

**Device:** Moto G5 Plus
**Android version:** 8.1.0
**Signal version:** 4.15.5 

### Link to debug log

https://gist.github.com/ef19812b4618762bbdb20bc144f6f00e

**Comment by p4nci** [comment]
@rannbaron you see numbers instead of names? Generally [this](https://github.com/signalapp/Signal-Android/issues/4338#issuecomment-198157982) helps.
Btw, seems like a duplicate of https://github.com/signalapp/Signal-Android/issues/5079

**Comment by rannbaron** [comment]
I have deleted the account and resynced many times. Same thing: no contact names. This isn't a duplicate of the issue you mentioned. The errors in the debug log are very different.

**Comment by p4nci** [comment]
@rannbaron I had a similar issue and for me it worked ONLY after a reboot and/or "kill Signal" (sorry, I don't remember now). Probably it was a different issue 😞

**Comment by rannbaron** [comment]
This is the relevant section:

02-05 10:49:34.316   943  1230 W ContactsDatabase: Adding number: +*********99
02-05 10:49:34.331   943  1230 W ContactsDatabase: Adding number: +*********97

<lots more of these>

02-05 10:49:34.434   943  1230 E JavaBinder: !!! FAILED BINDER TRANSACTION !!!  (parcel size = 1212188)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 
02-05 10:49:34.444   943  1230 W DirectoryHelper: android.os.TransactionTooLargeException: data parcel size 1212188 bytes
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at android.os.BinderProxy.transactNative(Native Method)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at android.os.BinderProxy.transact(Binder.java:764)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at android.content.ContentProviderProxy.applyBatch(ContentProviderNative.java:520)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at android.content.ContentProviderClient.applyBatch(ContentProviderClient.java:465)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at android.content.ContentResolver.applyBatch(ContentResolver.java:1578)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at org.thoughtcrime.securesms.contacts.ContactsDatabase.setRegisteredUsers(ContactsDatabase.java:121)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at org.thoughtcrime.securesms.util.DirectoryHelper.updateContactsDatabase(DirectoryHelper.java:163)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at org.thoughtcrime.securesms.util.DirectoryHelper.refreshDirectory(DirectoryHelper.java:111)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at org.thoughtcrime.securesms.util.DirectoryHelper.refreshDirectory(DirectoryHelper.java:59)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at org.thoughtcrime.securesms.ContactSelectionActivity$RefreshDirectoryTask.doInBackground(ContactSelectionActivity.java:143)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at org.thoughtcrime.securesms.ContactSelectionActivity$RefreshDirectoryTask.doInBackground(ContactSelectionActivity.java:128)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at android.os.AsyncTask$2.call(AsyncTask.java:333)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
02-05 10:49:34.444   943  1230 W DirectoryHelper: 	at java.lang.Thread.run(Thread.java:764)

**Comment by rannbaron** [comment]
@moxie0 I fixed this by moving the lines

      if (!operations.isEmpty()) {
        context.getContentResolver().applyBatch(ContactsContract.AUTHORITY, operations);
      }

in ContactsDatabase.java to within the loop above them (diff below). This clearly isn't a good solution, as it adds each contact to the database separately, but it does solve the problem. As the error above shows, the issue is that for some reason, adding all the contacts at once creates too big a transaction.

diff --git a/src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java b/src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java
index 34c0e6f6..7078ae49 100644
--- a/src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java
+++ b/src/org/thoughtcrime/securesms/contacts/ContactsDatabase.java
@@ -115,11 +115,12 @@ public class ContactsDatabase {
         Log.w(TAG, "Updating display name: " + currentContactEntry.getKey());
         updateDisplayName(operations, currentContactEntry.getValue().getAggregateDisplayName(), currentContactEntry.getValue().getId(), currentContactEntry.getValue().getDisplayNameSource());
       }
+      if (!operations.isEmpty()) {
+        context.getContentResolver().applyBatch(ContactsContract.AUTHORITY, operations);
+      }
     }
 
-    if (!operations.isEmpty()) {
-      context.getContentResolver().applyBatch(ContactsContract.AUTHORITY, operations);
-    }
+
   }
 
   @NonNull Cursor querySystemContacts(@Nullable String filter) {

**Comment by moxie0** [comment]
Thanks @rannbaron, that's a really well researched report! I wonder if you have contacts with really long display names or just a lot of Signal contacts?

**Comment by rannbaron** [comment]
@moxie0 Thank you! I just grabbed my contacts2.db file (from com.android.providers.contacts), opened it in sqlitebrowser and copied the display names, then ran the resulting text file through a Python script to count them and check their length. The longest is 39 characters. I have 1,470 contact display names altogether. In case it matters, there were a few NULL entries in the table as well.

**Comment by moxie0** [comment]
And can you see about how many of those are Signal users?

**Comment by rannbaron** [comment]
Not in this database, but it's probably not more than 25%.

**Comment by moxie0** [comment]
I can apply the batched operations more frequently in that loop, but I'm trying to figure out how frequently. I think calculating the serialized length of each operation would be slightly expensive, and I'm trying to understand why this is affecting you but hasn't really appeared to affect anyone else so far. 

It sounds like you have a not insane number of contacts, and that those contacts don't have insane display names.

**Comment by rannbaron** [comment]
I agree! This is quite odd. I will say that this arose after I upgraded my LineageOS rom to a new version that strictly enforces SELinux. I had backed up my old Signal data with Titanium Backup, then restored it. That crashed Signal, so I cleared all data. When I created a new account, this started happening. It was not happening before the upgrade. But then, I had not reinstalled Signal in a very long time, so it never needed to resync my entire database at once.

No idea if this is useful...I'm happy to help out any way I can!

**Comment by f4stb00t** [comment]
Hi folks!

First of all thanks a lot to moxie for that outstanding app. Keep up the good work! 

Actually I am facing the same issues. I cannot see any contacts. I thought it's because of all the weekly updates I've installed on my phone, but It's not. 

Some facts:
Android: lineage-14.1-20180222-nightly-m8-signed.zip
Signal: v4.15.5
Bug Reproducable: yes
Root: yes
GAPPS: nope

What has been done to fix the issue?
- Refresh Contactlist: NOK
- delete signal account and readd: NOK
- purge and reinstall signal: NOK
- remove and readd contacts: NOK
- remove all contacts and add only one by hand (signal user): NOK

Therefore I tried  to fix it with a fresh install of LineageOS
- Install newest los
- root device
- import my contactlist backup
- install fdroid, afwall
- install signal apk(also tried with yalp version)
Result: It's showing up the contacts again. But when I reboot the device all contacts disappeared.  Only the ones left where i had already a conversation open.
Because of this I played a bit with the settings. When I configure Signal to be used as sms app as well, I can see all contacts again, but even my Signal friends have only a phonenumber. Weird is that a bunch of numbers have no name in the signal list. But in fact they have a Name in my address book.

Would be great if you can give this a shot to fix that issue. 

If I can provide you any info you need. Please let me know. Just with the debug log will be hard. Since I installed everything again. Don't want to reboot the device again to lose all my contacts :/

Cheers

Fastb00t

**Comment by joaobarcia** [comment]
Happening in one of my devices as well.

Model: MotoG3
Android: lineage-14.1-20180324-NIGTHLY-osprey
Android version: 7.1.2
Signal: v4.17.5
Bug Reproduceable: yes
Root: nope
GAPPS: nope

Tried:
- Updating Lineage
- Uninstall > Re-install Signal
- Explicitly setting "read contacts" and "modify contacts" to "Allowed"

Thanks

**Comment by automated-signal** [comment]
GitHub Issue Cleanup:
See #7598 for more information.
